### ISSUE TITLE
Mangling file names makes it hard to track down errors

### ISSUE THREAD

**Issue by deliciouslytyped** [theam/require#17]
```
/run/user/1000/ghc20989_0/ghc_13.hspp:35:5: error:
    Not in scope: ‘Widget.registerCallback’
    Module ‘GUI.Gallery.Widget’ does not export ‘registerCallback’.
   |
35 |     Widget.registerCallback widget (Widget.UserEvent userEventCallback)
   |     ^^^^^^^^^^^^^^^^^^^^^^^

```

**Comment by NickSeagull** [comment]
Hi! Thanks for reporting this 😄 

Could you provide a little bit more of detail?
* What are you trying to achieve?
* What's the problematic code?
* What is the expected behavior?

**Comment by deliciouslytyped** [comment]
Hey, sorry for the late reply, I don't have access to the original stuff right now but unless I missed something, my problem was that it's not clear from `/run/user/1000/ghc20989_0/ghc_13.hspp:35:5:` which file the error originates from.

**Comment by NickSeagull** [comment]
Oh yeah, this is a current limitation given that we are using GHC's preprocessing capabilities. We don't do the mangling ourselves, I have to try whether writing to our own filename works or not, but I'm not very optimistic on that :sweat_smile: 

If you have any kind of idea, feel free to comment about it, I'd love to solve this.

**Comment by deliciouslytyped** [comment]
I've never looked at anything like this, sorry! Maybe the GHC people might have suggestions?

**Comment by NickSeagull** [comment]
Hmm, actually I might ask :)

**Comment by deliciouslytyped** [comment]
Any hope for this?

**Comment by NickSeagull** [comment]
Hey thanks for reaching out!

After researching this for some time, the only way that GHC would allow to give proper paths is by forking it and creating a custom extension. The issue with this, is that it comes with many other "friends" like maintaining a custom fork, integrating into Cabal and Stack, etc...

Also, we cannot generate the code from Template Haskell, because it [disallows import generation](https://gitlab.haskell.org//ghc/ghc/issues/1475#note_84043).

There is a possible solution that came to my mind, which is storing these intermediate files in a `.require` folder, so one can inspect them. What do you think about this solution @deliciouslytyped ?

**Comment by deliciouslytyped** [comment]
It sounds non-ideal, but feasible. I don't know. Would it be reasonable to file a bug against the API or a feature request? Sounds like this could be an issue for other preprocessors as well?

I think the template haskell limitation is one of the reasons I ended up here.

**Comment by NickSeagull** [comment]
Maybe, probably it would be a matter of asking in the `#ghc` channel on the FreeNode IRC

**Comment by bgamari** [comment]
Indeed it does look like this is a GHC bug given that we are using the name of a (presumably GHC-generated) temporary file in an error message. Can you open a GHC ticket with a reproducer, @deliciouslytyped ?

**Comment by deliciouslytyped** [comment]
I don't know how to reproduce this anymore.

**Comment by deliciouslytyped** [comment]
Here is a reproducer, if you use Nix you can just run `repro.sh` and it will use the haskell.nix to run everything.
[repro.zip](https://github.com/theam/require/files/4430677/repro.zip)

**Comment by deliciouslytyped** [comment]
This may be related? https://gitlab.haskell.org/ghc/ghc/issues/13266

**Comment by deliciouslytyped** [comment]
@NickSeagull please see https://gitlab.haskell.org/ghc/ghc/issues/18017#note_263706

**Comment by NickSeagull** [comment]
Just took a look, thanks for pointing this out!

**Comment by NickSeagull** [comment]
Should be fixed in `0.4.5`. Feel free to reopen if the issue persists :)
